snapcraft/storeapi/__init__.py

```diff
@@ -13,9 +13,282 @@
+        # Will be set by download()
+        self.cpi = None
+
+    # API for snapcraft
+
+    def login(self, email, password, one_time_password=None):
+        result = dict(success=False, body=None)
+        acls = ('store_read', 'store_write',
+                'package_access', 'package_upload')
+        macaroons = {}
+        for acl in acls:
+            macaroon, error = self.get_macaroon(acl)
+            if error:
+                result['errors'] = error
+                break
+            macaroons[acl] = macaroon
+        else:
+            discharges, error = self.get_discharges(
+                email, password, one_time_password, macaroons)
+            if error:
+                result['errors'] = error
+            else:
+                # All macaroons have been discharged, save them in the config
+                for acl in acls:
+                    self.conf.set(acl, ','.join([macaroons[acl],
+                                                 discharges[acl]]))
+                self.conf.save()
+                result['success'] = True
+        return result
+
+    def logout(self):
+        self.conf.clear()
+        self.conf.save()
+
+    def register_name(self, name):
+        data = dict(snap_name=name, series=DEFAULT_SERIES)
+        auth = _macaroon_auth(self.conf, 'package_upload')
+        response = self.post('register-name/',
+                             data=json.dumps(data),
+                             headers={'Authorization': auth,
+                                      'Content-Type': 'application/json'})
+        if not response.ok:
+            # if (response['errors'] == ['Authorization Required']
+            #     and (response.headers['WWW-Authenticate']
+            #          == "Macaroon needs_refresh=1":
+            # Refresh the discharge macaroon and retry
+            pass
+        return response
+
+    def upload(self, binary_path, snap_name):
+        if self.conf.get('package_upload') is None:
+            raise InvalidCredentialsError()
+        self.updown = requests.Session()
+        data = _upload.upload_files(binary_path, self.updown)
+        success = data.get('success', False)
+        if not success:
+            return data
+
+        result = _upload.upload_app(self, snap_name, data)
+        return result
+
+    def download(self, snap_name, channel, download_path, arch=None):
+        if arch is None:
+            arch = snapcraft.ProjectOptions().deb_arch
+        self.cpi = CPIClient(self.conf)
+        logger.info('Getting details for {}'.format(snap_name))
+        package = self.cpi.search_package(snap_name, channel, arch)
+        if package is None:
+            raise SnapNotFoundError(snap_name, channel, arch)
+        return self.download_snap(snap_name, channel, arch,
+                                  download_path, package['download_url'],
+                                  package['download_sha512'])
+
+    def get_macaroon(self, acl):
```

Comment:
even on files with _something.py, I find it really useful to mark the functions that are only called from inside that file with a _. I prefer to use _func until it's clear that it will be needed outside, not the other way around.